### DISCUSSION TITLE
디바이스에 데이터를 저장하는 방법은 SwiftData만 있는 것이 아닙니다!

### DISCUSSION THREAD

**Discussion by chongin12** [DeveloperAcademy-POSTECH/.github#14] [Idea]
안녕하세요! 3기 오전 세션 러너 Mosu입니다~!!

디바이스에 값을 저장할 때 UserDefaults도 사용할 수 있습니다.
SwiftData처럼 모델간 관계 정의나 기타 편의 기능 등을 사용하진 못하지만, 단순하게 데이터를 저장하기에는 이게 더 편합니다!
 
SwiftUI에서는 `@AppStorage` 라는 프로퍼티 래퍼를 사용해서 더더욱 쉽게 사용할 수 있습니다.
기존에 사용하던 `@State` 자리에 `@AppStorage("키 이름")` 를 넣어주시면 바로 사용하실 수 있습니다.
 
```swift
struct ContentView: View {
    @AppStorage("isOn") var isOn: Bool = false
    var body: some View {
        Toggle("토글", isOn: $isOn)
    }
}
```
 
(내부적으로 .plist 파일에 해당 내용을 기록하기 때문에 추출이 매우 쉽습니다. 개인정보를 담지 않도록 주의하세요!)
 
그럼 모두 NC1 화이팅입니다!!

**Comment by moonkey48** [comment]
정리 감사합니다 모수!ㅎㅎ
추가적으로 온보딩과 같이 앱을 깔았을 때 처음 보여지는 화면을 보여지는 경우에도 많이 사용하게 되는 것 같아요ㅎㅎ
```swift
struct ContentView: View {
    @AppStorage("isOnboarding") var isOnboarding: Bool = false
    var body: some View {
        VStack {
          Text("온보딩입니다:)")
          Button {
            isOnboarding = false
          } label: {
            Text("온보딩 종료")
          }
        }
    }
}
```

**Comment by lingostar** [comment]
첨언하자면, 
.plist 는 XML 파일이기 때문에, 아카이빙이 아니라 시리얼라이제이션이라서 오브젝트 그래프를 저장할 수는 없습니다.
그리고 이미지 같은 바이너리 저장은 거의 불가능해요. (.plist는 1MB 정도의 용량 제한이 있는 걸로 알고 있어요)
파일이 커지면 엄청 느려지기도 한답니다.

결론적으로, User Defaults 같이 환경설정을 위해서는 아주 간편한 방법이지만, 계속 쌓여나가는 사용자의 데이터를 저장하기는 부적합합니다.

**Reply by Jaehwa-Noh** [reply]
AppStorage는 User Defaults랑 동일한 것으로 알고 있어요.
https://developer.apple.com/documentation/swiftui/appstorage
